contracts/PvpArena.sol

```diff
@@ -728,124 +688,214 @@ contract PvpArena is Initializable, AccessControlUpgradeable {
+        return prizePercentages;
     }
 
-    /// @dev wether or not the character is the attacker in a duel
-    // and has not performed an action
-    function hasPendingDuel(uint256 characterID) public view returns (bool) {
-        return duelByAttacker[characterID].isPending;
+    /// @dev returns the account's ranking prize pool earnings
+    function getPlayerPrizePoolRewards() public view returns (uint256) {
+        return _rankingRewardsByPlayer[msg.sender];
     }
 
-    /// @dev wether or not a character can appear as someone's opponent
-    function isCharacterAttackable(uint256 characterID)
-        public
-        view
-        returns (bool)
-    {
-        uint256 lastActivity = _lastActivityByCharacter[characterID];
+    /// @dev returns the current duel queue
+    function getDuelQueue() public view returns (uint256[] memory) {
+        uint256 length = _duelQueue.length();
+        uint256[] memory values = new uint256[](length);
 
-        return lastActivity.add(unattackableSeconds) <= block.timestamp && !_duelQueue.contains(characterID) && !characterDefending[characterID];    
+        for (uint256 i = 0; i < length; i++) {
+            values[i] = _duelQueue.at(i);
+        }
+
+        return values;
     }
 
-    /// @dev updates the last activity timestamp of a character
-    function _updateLastActivityTimestamp(uint256 characterID) private {
-        _lastActivityByCharacter[characterID] = block.timestamp;
+    /// @dev assigns an opponent to a character
+    function _assignOpponent(uint256 characterID) private {
+        uint8 tier = getArenaTier(characterID);
+        EnumerableSet.UintSet
+            storage matchableCharacters = _matchableCharactersByTier[tier];
+
+        require(matchableCharacters.length() != 0, "No enemy in tier");
+        require(!_duelQueue.contains(characterID), "Char dueling");
+
+        uint256 seed = randoms.getRandomSeed(msg.sender);
+        uint256 randomIndex = RandomUtil.randomSeededMinMax(
+            0,
+            matchableCharacters.length() - 1,
+            seed
+        );
+        uint256 opponentID;
+        uint256 matchableCharactersCount = matchableCharacters.length();
+        bool foundOpponent = false;
+
+        for (uint256 i = 0; i < matchableCharactersCount; i++) {
+            uint256 index = (randomIndex + 1) % matchableCharactersCount;
+            uint256 candidateID = matchableCharacters.at(index);
+
+            if (candidateID == characterID) {
+                if (matchableCharactersCount == 1) {
+                    break;
+                }
+                if (
+                    matchableCharacters.at(matchableCharactersCount - 1) ==
+                    candidateID
+                ) {
+                    candidateID = matchableCharacters.at(0);
+                } else {
+                    candidateID = matchableCharacters.at(index + 1);
+                }
+            }
+            if (
```

Comment:
This logic appears to be preventing the characters in the same account from competing with each other in PVP.

Recommend:
`_matchableCharactersByTier` should change from:
`mapping(uint8 => EnumerableSet.UintSet) private _matchableCharactersByTier;`
to
`mapping(uint8 => mapping(address => EnumerableSet.UintSet)) private _matchableCharactersByTier;`

Then the for loop disappears entirely.  You pick a random index.  That index indicates an account with 1-4 characters ready to be matched.  If the index picks the current character, then use the next slot (with wraparound) which is guaranteed to be a different account.

Note that the explicit check `if (matchableCharactersCount == 1)` may also become unnecessary.

Reply:
Sorry, proposed data structure was wrong above.

```
struct MatchableAccount {
    address owner;
    EnumerableSet.UintSet characters;
}

mapping(uint8 => MatchableAccount) _matchableAccountsByTier;
```

Existing
- `uint256 candidateID = matchableCharacters.at(index);`
- check in the for loop to see if account is matching and continue loop if so

Proposed
- `MatchableAccount matchableAccount = matchableAccounts.at(index);`
- check `matchableAccount.owner` for match, if matched fetch `matchableAccount = matchableAccounts[(index+1)%matchableAccountsCount]`
- you can assert that the 2nd matchableAccount owner is different
- no loop